File: resharper/test/data/References/InvalidMemberDataProperties.vb
Imports System
Imports System.Collections.Generic
Imports Xunit

Public Class MyTest
    <Theory>
    <MemberData("NonPublicDataEnumerator")>
    Public Sub Test1(ByVal value As Int32)
        Assert.Equal(42, value)
    End Sub

    <Theory>
    <MemberData("NonStaticDataEnumerator")>
    Public Sub Test1(ByVal value As Int32)
        Assert.Equal(42, value)
    End Sub

    <Theory>
    <MemberData("InvalidReturnTypeDataEnumerator")>
    Public Sub Test1(ByVal value As Int32)
        Assert.Equal(42, value)
    End Sub

    <Theory>
    <MemberData("NonPublicMethod")>
    Public Sub Test1(ByVal value As Int32)
        Assert.Equal(42, value)
    End Sub

    <Theory>
    <MemberData("NonStaticMethod")>
    Public Sub Test1(ByVal value As Int32)
        Assert.Equal(42, value)
    End Sub

    <Theory>
    <MemberData("InvalidReturnTypeMethod")>
    Public Sub Test1(ByVal value As Int32)
        Assert.Equal(42, value)
    End Sub

    <Theory>
    <MemberData("NonPublicField")>
    Public Sub Test1(ByVal value As Int32)
        Assert.Equal(42, value)
    End Sub

    <Theory>
    <MemberData("NonStaticField")>
    Public Sub Test1(ByVal value As Int32)
        Assert.Equal(42, value)
    End Sub

    <Theory>
    <MemberData("InvalidReturnTypeField")>
    Public Sub Test1(ByVal value As Int32)
        Assert.Equal(42, value)
    End Sub

    Private Shared ReadOnly Iterator Property NonPublicDataEnumerator() As IEnumerable(Of Object())
        Get
            Yield New Object() { 42 }
        End Get
    End Property

    Public ReadOnly Iterator Property NonStaticDataEnumerator() As IEnumerable(Of Object())
        Get
            Yield New Object() { 42 }
        End Get
    End Property

    Public Shared ReadOnly Iterator Property InvalidReturnTypeDataEnumerator() As IEnumerable(Of Int32())
        Get
            Yield New Int32() { 42 }
        End Get
    End Property

    Private Shared Iterator Function NonPublicMethod() As IEnumerable(Of Object())
        Yield New Object() { 42 }
    End Function

    Public Iterator Function NonStaticMethod() As IEnumerable(Of Object())
        Yield New Object() { 42 }
    End Function

    Public Shared Iterator Function InvalidReturnTypeMethod() As IEnumerable(Of Int32())
        Yield New Int32() { 42 }
    End Function

    Private Shared NonPublicField As IEnumerable(Of Object())
    Public NonStaticField As IEnumerable(Of Object())
    Public Shared InvalidReturnTypeField As IEnumerable(Of Int32())
End Class

File: resharper/test/data/References/InvalidMemberDataProperties.cs
using System.Collections.Generic;
using Xunit;

namespace Foo
{
    public class MyTest
    {
        [Theory]
        [MemberData("NonPublicDataEnumerator")]
        public void Test1(int value)
        {
            Assert.Equal(42, value);
        }

        [Theory]
        [MemberData("NonStaticDataEnumerator")]
        public void Test2(int value)
        {
            Assert.Equal(42, value);
        }

        [Theory]
        [MemberData("InvalidReturnTypeDataEnumerator")]
        public void Test3(int value)
        {
            Assert.Equal(42, value);
        }

        [Theory]
        [MemberData("NonPublicMethod")]
        public void Test4(int value)
        {
            Assert.Equal(42, value);
        }

        [Theory]
        [MemberData("NonStaticMethod")]
        public void Test5(int value)
        {
            Assert.Equal(42, value);
        }

        [Theory]
        [MemberData("InvalidReturnTypeMethod")]
        public void Test6(int value)
        {
            Assert.Equal(42, value);
        }

        [Theory]
        [MemberData("NonPublicField")]
        public void Test7(int value)
        {
            Assert.Equal(42, value);
        }

        [Theory]
        [MemberData("NonStaticField")]
        public void Test8(int value)
        {
            Assert.Equal(42, value);
        }

        [Theory]
        [MemberData("InvalidReturnTypeField")]
        public void Test9(int value)
        {
            Assert.Equal(42, value);
        }

        // Needs to be public
        private static IEnumerable<object[]> NonPublicDataEnumerator
        {
            get { yield return new object[] { 42 }; }
        }

        // Needs to be static
        public IEnumerable<object[]> NonStaticDataEnumerator
        {
            get { yield return new object[] { 42 }; }
        }

        // Invalid return type signature
        public static IEnumerable<int[]> InvalidReturnTypeDataEnumerator
        {
            get { return new List<int[]>(new int[] { 42 }); }
        }
 
        // Needs to be public
        private static IEnumerable<object[]> NonPublicMethod()
        {
            yield return new object[] { 42 };
        }

        // Needs to be static
        public IEnumerable<object[]> NonStaticMethod()
        {
            yield return new object[] { 42 };
        }

        // Invalid return type signature
        public static IEnumerable<int[]> InvalidReturnTypeMethod()
        {
            return new List<int[]>(new int[] { 42 });
        }
 
        // Needs to be public
        private static IEnumerable<object[]> NonPublicField = new List<object[]> {
            new object[] { 42 }
        };

        // Needs to be static
        public IEnumerable<object[]> NonStaticField = new List<object[]> {
            new object[] { 42 }
        };

        // Invalid return type signature
        public static IEnumerable<int[]> InvalidReturnTypeField = new List<int[]> {
            new int[] { 42 }
        };
    }
}
